### PR TITLE
fix: avoid reloading all csses when hot load

### PR DESCRIPTION
**Pull request by yiminghe**
This PR contains a:

- [ ] **bugfix**
- [ ] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [x] **metadata update**

### Motivation / Use-Case
- only refresh page when locals change, otherwise accept and hot load
- avoid reloading all csses 

### Additional Info

Fixes #692

### CODE REVIEW

File: src/loader.js [hunk 1]
```diff
@@ -44,9 +44,10 @@ const MiniCssExtractPlugin = require("./index");
  * @returns {string}
  */
 function hotLoader(content, context) {
-  const accept = context.locals
+  const hasLocals = !!(context.locals && Object.keys(context.locals).length);
+  const accept = hasLocals
     ? ""
-    : "module.hot.accept(undefined, cssReload);";
```
Comment (alexander-akait) [inline src/loader.js:49]: Why you remove it?

**Review by alexander-akait — COMMENTED** [review]
We need a test case, please add

File: src/loader.js (hunk 1 above)
Comment (yiminghe) [reply src/loader.js:49]: self accept is enough, it will update when dispose: https://webpack.js.org/api/hot-module-replacement/

**Comment by yiminghe** [comment]
> We need a test case, please add

test cases updated

File: src/loader.js [hunk 2]
```diff
@@ -44,22 +44,31 @@ const MiniCssExtractPlugin = require("./index");
...
-    ? ""
-    : "module.hot.accept(undefined, cssReload);";
-
+  const localsJsonString = JSON.stringify(JSON.stringify(context.locals));
   return `${content}
     if(module.hot) {
+      const localsJsonString = ${localsJsonString};
       // ${Date.now()}
-      var cssReload = require(${stringifyRequest(
+      const cssReload = require(${stringifyRequest(
```
Comment (alexander-akait) [inline src/loader.js:52]: Don't use const, it will be in the browser and break old browsers

File: src/loader.js [hunk 3]
```diff
@@ -44,22 +44,31 @@ const MiniCssExtractPlugin = require("./index");
  * @returns {string}
  */
 function hotLoader(content, context) {
-  const accept = context.locals
-    ? ""
-    : "module.hot.accept(undefined, cssReload);";
-
+  const localsJsonString = JSON.stringify(JSON.stringify(context.locals));
```
Comment (alexander-akait) [inline src/loader.js:47]: Why double `JSON.stringify`? Just typo?

File: test/cases/hmr/expected/main.js [hunk 4]
```diff
@@ -363,10 +363,23 @@ __webpack_require__.r(__webpack_exports__);
...
+        module.hot.data.value !== localsJsonString
+      ) {
+        module.hot.invalidate();
+      } else {
+        module.hot.accept();
+      }
+      module.hot.dispose(function(data) {
+        data.value = localsJsonString;
+        cssReload();
+      });
```
Comment (alexander-akait) [inline test/cases/hmr/expected/main.js:382]: We have a special test directory for HMR, will be great to add a test case there, I know your code works fine, but we should ensure about it

**Review by alexander-akait — CHANGES_REQUESTED** [review]
Let's do small improvements and I am fine with it

**Comment by alexander-akait** [comment]
For fixing tests for OLD API just use `OLD_API=1 npm run test:only`

**Comment by alexander-akait** [comment]
Thank you for your PR

File: src/loader.js (hunk 3 above)
Comment (yiminghe) [reply src/loader.js:47]: `const localsJsonString = ${localsJsonString};`

string comparation is faster

File: test/cases/hmr/expected/main.js (hunk 4 above)
Comment (yiminghe) [reply test/cases/hmr/expected/main.js:382]: what do you mean? I have added cases inside cases folder (hmr, hmr-locals)

**Comment by yiminghe** [comment]
> For fixing tests for OLD API just use `OLD_API=1 npm run test:only`

passed

File: test/cases/hmr/expected/main.js (hunk 4 above)
Comment (alexander-akait) [reply test/cases/hmr/expected/main.js]: Sorry for delay, just put a test case here https://github.com/webpack-contrib/mini-css-extract-plugin/blob/master/test/HMR.test.js

Comment (yiminghe) [reply test/cases/hmr/expected/main.js]: > Sorry for delay, just put a test case here https://github.com/webpack-contrib/mini-css-extract-plugin/blob/master/test/HMR.test.js

done

**Comment by alexander-akait** [comment]
@yiminghe We have:
```
var cssReload = require("/home/runner/work/mini-css-extract-plugin/mini-css-extract-plugin/src/hmr/hotModuleReplacement.js")(module.id, undefined);
```

As you can see it is an absolute path, let's use like here https://github.com/webpack-contrib/css-loader/blob/master/src/utils.js#L15

**Comment by yiminghe** [comment]
> @yiminghe We have:
> 
> ```
> var cssReload = require("/home/runner/work/mini-css-extract-plugin/mini-css-extract-plugin/src/hmr/hotModuleReplacement.js")(module.id, undefined);
> ```
> 
> As you can see it is an absolute path, let's use it from here https://github.com/webpack-contrib/css-loader/blob/master/src/utils.js#L15 (maybe we already have the same util here)

updated

**Review by alexander-akait — DISMISSED** [review]

**Pull request merged**